File: output/test_project/calculator/calculator.py
#!/usr/bin/env python3
"""
Advanced calculator implementation with various mathematical operations.
"""
import math
import re
from typing import Union, List, Tuple, Dict, Optional

def hello_world():
    """Print a greeting message"""
    print("Hello, world!")

def add_numbers(a: Union[int, float], b: Union[int, float]) -> Union[int, float]:
    """Add two numbers together
    
    Args:
        a: First number
        b: Second number
        
    Returns:
        Sum of a and b
    """
    return a + b

def parse_expression(expression: str) -> Union[float, str]:
    """Parse and evaluate a simple mathematical expression
    
    Args:
        expression: Mathematical expression as string (e.g., "2 + 3 * 4")
        
    Returns:
        Result of the evaluated expression
        
    Raises:
        ValueError: If the expression is invalid or contains unsupported operations
        ZeroDivisionError: If the expression attempts division by zero
    """
    # Security check - only allow basic arithmetic operations and numbers
    if not re.match(r'^[\d\s\+\-\*\/\(\)\.]+$', expression):
        raise ValueError("Expression contains invalid characters")
    
    try:
        # Replace ^ with ** for exponentiation
        expression = expression.replace('^', '**')
        # Evaluate the expression safely with limited math functions
        safe_dict = {"__builtins__": None}
        safe_dict.update({
            "abs": abs,
            "round": round,
            "max": max,
            "min": min
        })
        result = eval(expression, {"__builtins__": None}, safe_dict)
        return result
    except ZeroDivisionError:
        raise ZeroDivisionError("Cannot divide by zero")
    except Exception as e:
        raise ValueError(f"Invalid expression: {e}")

class Calculator:
    """An advanced calculator class with memory and history"""
    
    def __init__(self, initial_value: Union[int, float] = 0):
        """Initialize the calculator
        
        Args:
            initial_value: Initial value for the calculator
        """
        self.result = initial_value
        self.memory = 0.0
        self.history: List[Dict[str, Union[str, float]]] = []
        self._record_operation("init", initial_value, initial_value)
    
    def _record_operation(self, operation: str, value: Union[int, float], result: Union[int, float]) -> None:
        """Record an operation in the history
        
        Args:
            operation: The operation performed
            value: The value used in the operation
            result: The result after the operation
        """
        self.history.append({
            "operation": operation,
            "value": value,
            "result": result
        })
    
    def add(self, number: Union[int, float]) -> Union[int, float]:
        """Add a number to the result
        
        Args:
            number: Number to add
            
        Returns:
            Updated result
        """
        self.result += number
        self._record_operation("add", number, self.result)
        return self.result
    
    def subtract(self, number: Union[int, float]) -> Union[int, float]:
        """Subtract a number from the result
        
        Args:
            number: Number to subtract
            
        Returns:
            Updated result
        """
        self.result -= number
        self._record_operation("subtract", number, self.result)
        return self.result
    
    def multiply(self, number: Union[int, float]) -> Union[int, float]:
        """Multiply the result by a number
        
        Args:
            number: Number to multiply by
            
        Returns:
            Updated result
        """
        self.result *= number
        self._record_operation("multiply", number, self.result)
        return self.result
    
    def divide(self, number: Union[int, float]) -> Union[int, float]:
        """Divide the result by a number
        
        Args:
            number: Number to divide by
            
        Returns:
            Updated result
            
        Raises:
            ZeroDivisionError: If number is 0
        """
        if number == 0:
            raise ZeroDivisionError("Cannot divide by zero")
        self.result /= number
        self._record_operation("divide", number, self.result)
        return self.result
    
    def power(self, exponent: Union[int, float]) -> Union[int, float]:
        """Raise the result to a power
        
        Args:
            exponent: The power to raise to
            
        Returns:
            Updated result
        """
        self.result **= exponent
        self._record_operation("power", exponent, self.result)
        return self.result
    
    def square_root(self) -> Union[int, float]:
        """Calculate the square root of the result
        
        Returns:
            Updated result
            
        Raises:
            ValueError: If result is negative
        """
        if self.result < 0:
            raise ValueError("Cannot calculate square root of a negative number")
        self.result = math.sqrt(self.result)
        self._record_operation("sqrt", None, self.result)
        return self.result
    
    def memory_store(self) -> None:
        """Store the current result in memory"""
        self.memory = self.result
        self._record_operation("m_store", self.result, self.result)
    
    def memory_recall(self) -> Union[int, float]:
        """Recall the value from memory
        
        Returns:
            The memory value
        """
        self._record_operation("m_recall", self.memory, self.result)
        return self.memory
    
    def memory_add(self) -> None:
        """Add the current result to memory"""
        self.memory += self.result
        self._record_operation("m_add", self.result, self.result)
    
    def memory_clear(self) -> None:
        """Clear the memory"""
        self.memory = 0.0
        self._record_operation("m_clear", 0, self.result)
    
    def clear(self) -> None:
        """Clear the result"""
        self.result = 0.0
        self._record_operation("clear", 0, self.result)
    
    def get_history(self) -> List[Dict[str, Union[str, float]]]:
        """Get the operation history
        
        Returns:
            List of operations performed
        """
        return self.history
    
    def evaluate(self, expression: str) -> Union[int, float]:
        """Evaluate a mathematical expression and update the result
        
        Args:
            expression: The expression to evaluate
            
        Returns:
            The result after evaluation
            
        Raises:
            ValueError: If expression is invalid
        """
        try:
            value = parse_expression(expression)
            self.result = value
            self._record_operation("evaluate", expression, value)
            return value
        except (ValueError, ZeroDivisionError) as e:
            self._record_operation("error", expression, str(e))
            raise

File: output/test_project/web_app/src/utils/calculator.ts
/**
 * Calculator utility functions for the web application.
 * Provides arithmetic operations and expression parsing.
 */

export interface CalculationResult {
  result: number;
  expression?: string;
  operation?: string;
  operands?: number[];
  timestamp: Date;
}

export interface CalculatorMemory {
  value: number;
  history: CalculationResult[];
  lastOperation?: string;
}

/**
 * Calculator class that provides arithmetic operations and memory functions.
 */
export class Calculator {
  private memory: CalculatorMemory = {
    value: 0,
    history: [],
  };

  /**
   * Adds two or more numbers together.
   * @param numbers - The numbers to add
   * @returns The result of the addition
   */
  add(...numbers: number[]): CalculationResult {
    if (numbers.length < 2) {
      throw new Error('At least two numbers are required for addition');
    }
    
    const result = numbers.reduce((sum, num) => sum + num, 0);
    
    const calculationResult: CalculationResult = {
      result,
      operation: 'add',
      operands: numbers,
      timestamp: new Date()
    };
    
    this.memory.history.push(calculationResult);
    this.memory.lastOperation = 'add';
    
    return calculationResult;
  }

  /**
   * Subtracts the second number from the first.
   * @param a - The first number
   * @param b - The number to subtract
   * @returns The result of the subtraction
   */
  subtract(a: number, b: number): CalculationResult {
    const result = a - b;
    
    const calculationResult: CalculationResult = {
      result,
      operation: 'subtract',
      operands: [a, b],
      timestamp: new Date()
    };
    
    this.memory.history.push(calculationResult);
    this.memory.lastOperation = 'subtract';
    
    return calculationResult;
  }

  /**
   * Multiplies two or more numbers together.
   * @param numbers - The numbers to multiply
   * @returns The result of the multiplication
   */
  multiply(...numbers: number[]): CalculationResult {
    if (numbers.length < 2) {
      throw new Error('At least two numbers are required for multiplication');
    }
    
    const result = numbers.reduce((product, num) => product * num, 1);
    
    const calculationResult: CalculationResult = {
      result,
      operation: 'multiply',
      operands: numbers,
      timestamp: new Date()
    };
    
    this.memory.history.push(calculationResult);
    this.memory.lastOperation = 'multiply';
    
    return calculationResult;
  }

  /**
   * Divides the first number by the second.
   * @param a - The dividend
   * @param b - The divisor
   * @returns The result of the division
   * @throws Error if divisor is zero
   */
  divide(a: number, b: number): CalculationResult {
    if (b === 0) {
      throw new Error('Division by zero is not allowed');
    }
    
    const result = a / b;
    
    const calculationResult: CalculationResult = {
      result,
      operation: 'divide',
      operands: [a, b],
      timestamp: new Date()
    };
    
    this.memory.history.push(calculationResult);
    this.memory.lastOperation = 'divide';
    
    return calculationResult;
  }

  /**
   * Raises a number to a power.
   * @param base - The base number
   * @param exponent - The exponent
   * @returns The result of the power operation
   */
  power(base: number, exponent: number): CalculationResult {
    const result = Math.pow(base, exponent);
    
    const calculationResult: CalculationResult = {
      result,
      operation: 'power',
      operands: [base, exponent],
      timestamp: new Date()
    };
    
    this.memory.history.push(calculationResult);
    this.memory.lastOperation = 'power';
    
    return calculationResult;
  }

  /**
   * Calculates the square root of a number.
   * @param number - The number to calculate the square root of
   * @returns The square root of the number
   * @throws Error if the number is negative
   */
  sqrt(number: number): CalculationResult {
    if (number < 0) {
      throw new Error('Cannot calculate square root of a negative number');
    }
    
    const result = Math.sqrt(number);
    
    const calculationResult: CalculationResult = {
      result,
      operation: 'sqrt',
      operands: [number],
      timestamp: new Date()
    };
    
    this.memory.history.push(calculationResult);
    this.memory.lastOperation = 'sqrt';
    
    return calculationResult;
  }

  /**
   * Evaluates a mathematical expression string.
   * @param expression - The expression to evaluate
   * @returns The result of the evaluation
   * @throws Error for invalid or unsafe expressions
   */
  evaluateExpression(expression: string): CalculationResult {
    // Clean and validate the expression
    const cleanExpression = expression.trim();
    
    // Validate that only allowed characters are used
    const validChars = /^[0-9\s\+\-\*\/\(\)\.\^\%]+$/;
    if (!validChars.test(cleanExpression)) {
      throw new Error('Expression contains invalid characters');
    }
    
    // Evaluate in a secure way using a sandboxed approach
    // For security reasons, we use a limited set of JavaScript math operations
    // rather than using eval() which could introduce security vulnerabilities
    
    // This is a simplified implementation for demonstration
    // In a real app, you would use a proper expression parser like math.js
    
    let result: number;
    try {
      // Convert ^ to Math.pow for exponentiation
      const preparedExpression = cleanExpression.replace(/(\d+|\))\s*\^\s*(\d+|\()/g, 'Math.pow($1, $2)');
      
      // Use Function constructor instead of eval for slightly better security
      // Still not 100% secure, but better than direct eval()
      const calculate = new Function(`
        'use strict';
        // Only allow these Math methods for security
        const Math = {
          pow: window.Math.pow,
          sqrt: window.Math.sqrt,
          abs: window.Math.abs,
          round: window.Math.round,
          floor: window.Math.floor,
          ceil: window.Math.ceil,
          sin: window.Math.sin,
          cos: window.Math.cos,
          tan: window.Math.tan,
          PI: window.Math.PI,
          E: window.Math.E
        };
        return ${preparedExpression};
      `);
      
      result = calculate();
    } catch (error) {
      throw new Error(`Failed to evaluate expression: ${error instanceof Error ? error.message : String(error)}`);
    }
    
    if (!Number.isFinite(result)) {
      throw new Error('Expression resulted in an invalid number');
    }
    
    const calculationResult: CalculationResult = {
      result,
      expression: cleanExpression,
      operation: 'evaluate',
      timestamp: new Date()
    };
    
    this.memory.history.push(calculationResult);
    this.memory.lastOperation = 'evaluate';
    
    return calculationResult;
  }
  
  /**
   * Stores a value in memory.
   * @param value - The value to store
   */
  memorySave(value: number): void {
    this.memory.value = value;
  }
  
  /**
   * Adds a value to the memory.
   * @param value - The value to add to memory
   */
  memoryAdd(value: number): void {
    this.memory.value += value;
  }
  
  /**
   * Subtracts a value from the memory.
   * @param value - The value to subtract from memory
   */
  memorySubtract(value: number): void {
    this.memory.value -= value;
  }
  
  /**
   * Retrieves the value stored in memory.
   * @returns The value stored in memory
   */
  memoryRecall(): number {
    return this.memory.value;
  }
  
  /**
   * Clears the memory value.
   */
  memoryClear(): void {
    this.memory.value = 0;
  }
  
  /**
   * Gets the calculation history.
   * @returns Array of past calculations
   */
  getHistory(): CalculationResult[] {
    return [...this.memory.history];
  }
  
  /**
   * Clears the calculation history.
   */
  clearHistory(): void {
    this.memory.history = [];
  }

  /**
   * Gets the last operation performed.
   * @returns The name of the last operation
   */
  getLastOperation(): string | undefined {
    return this.memory.lastOperation;
  }
}